lib/view_model/active_record/update_operation.rb

```diff
@@ -421,114 +455,173 @@ def build_updates_for_collection_association(association_data, association_updat
         clear_association_cache(model, association_data.direct_reflection)
       end
 
-      child_datas =
-        case association_update
-        when OwnedCollectionUpdate::Replace
-          association_update.update_datas
+      # Update contents are either UpdateData in the case of a nested
+      # association or reference strings in the case of a reference association.
+      # The former are resolved with resolve_child_viewmodels, the latter with
+      # resolve_referenced_viewmodels.
+      case association_update
+      when AbstractCollectionUpdate::Replace
+        child_datas = association_update.contents
 
-        when OwnedCollectionUpdate::Functional
-          child_datas =
-            previous_child_viewmodels.map do |previous_child_viewmodel|
-              UpdateData.empty_update_for(previous_child_viewmodel)
-            end
+      when AbstractCollectionUpdate::Functional
+        association_update.check_for_duplicates!(update_context, blame_reference)
 
-          association_update.check_for_duplicates!(update_context, self.viewmodel.blame_reference)
+        # Construct empty updates for previous children
+        child_datas =
+          previous_child_viewmodels.map do |previous_child_viewmodel|
+            UpdateData.empty_update_for(previous_child_viewmodel)
+          end
 
-          association_update.actions.each do |fupdate|
-            case fupdate
-            when FunctionalUpdate::Append
-              if fupdate.before || fupdate.after
-                moved_refs  = fupdate.contents.map(&:viewmodel_reference).to_set
-                child_datas = child_datas.reject { |child| moved_refs.include?(child.viewmodel_reference) }
+        # Insert or replace with either real UpdateData or reference strings
+        association_update.actions.each do |fupdate|
+          case fupdate
+          when FunctionalUpdate::Append
+            # If we're referring to existing members, ensure that they're removed before we append/insert
+            existing_refs = fupdate.contents
```

Comment:
I prefer the old name. I thought `existing_refs` was like `child_datas`, ie, the before state. The variable holds exactly the refs that are being moved by the current append action.